File: src/tui/theme.js
export const TITLE_ART = [
  "██╗   ██╗ ██████╗ ██████╗ ██╗  ██╗███████╗██████╗",
  "██║   ██║██╔═══██╗██╔══██╗██║ ██╔╝██╔════╝██╔══██╗",
  "██║   ██║██║   ██║██████╔╝█████╔╝ █████╗  ██████╔╝",
  "╚██╗ ██╔╝██║   ██║██╔══██╗██╔═██╗ ██╔══╝  ██╔══██╗",
  " ╚████╔╝ ╚██████╔╝██║  ██║██║  ██╗███████╗██║  ██║",
  "  ╚═══╝   ╚═════╝ ╚═╝  ╚═╝╚═╝  ╚═╝╚══════╝╚═╝  ╚═╝",
];

export const ANSI_PATTERN = /\x1b\[[0-9;]*m/g;

export const COLORS = {
  reset: "\x1b[0m",
  bold: "\x1b[1m",
  dim: "\x1b[2m",
  black: "\x1b[30m",
  green: "\x1b[32m",
  yellow: "\x1b[33m",
  red: "\x1b[31m",
  cyan: "\x1b[36m",
  magenta: "\x1b[35m",
  white: "\x1b[97m",
  gray: "\x1b[90m",
  brightGreen: "\x1b[92m",
  brightMagenta: "\x1b[95m",
  brightBlack: "\x1b[90m",
  bgGreen: "\x1b[42m",
  bgBrightGreen: "\x1b[102m",
  bgMagenta: "\x1b[45m",
  bgBrightMagenta: "\x1b[105m",
};

export function stripAnsi(text) {
  return String(text ?? "").replace(ANSI_PATTERN, "");
}

export function visibleLength(text) {
  return stripAnsi(text).length;
}

export function paint(text, styles, enabled) {
  if (!enabled || !styles || styles.length === 0) {
    return text;
  }

  const prefix = styles.map((style) => COLORS[style] ?? "").join("");
  return `${prefix}${text}${COLORS.reset}`;
}

export function colorize(text, color, enabled) {
  return paint(text, [color], enabled);
}

export function emphasize(text, enabled) {
  return paint(text, ["bold"], enabled);
}

export function highlight(text, enabled, options = {}) {
  if (!enabled) {
    return text;
  }

  return paint(text, ["bold", options.foreground ?? "black", options.background ?? "bgBrightGreen"], enabled);
}

export function truncate(text, maxLength) {
  const value = String(text ?? "");
  if (visibleLength(value) <= maxLength) {
    return value;
  }
  return `${stripAnsi(value).slice(0, Math.max(0, maxLength - 1))}…`;
}

export function pad(text, width) {
  const value = String(text ?? "");
  const missing = Math.max(0, width - visibleLength(value));
  return `${value}${" ".repeat(missing)}`;
}

export function fit(text, width) {
  return pad(truncate(text, width), width);
}

export function hardWrap(text, maxLength) {
  const value = stripAnsi(text);
  if (value.length <= maxLength) {
    return [value];
  }

  const lines = [];
  let remaining = value;

  while (remaining.length > maxLength) {
    const slice = remaining.slice(0, maxLength);
    const breakAt = slice.lastIndexOf(" ");
    if (breakAt >= Math.floor(maxLength * 0.6)) {
      lines.push(remaining.slice(0, breakAt));
      remaining = remaining.slice(breakAt + 1);
      continue;
    }

    lines.push(slice);
    remaining = remaining.slice(maxLength);
  }

  if (remaining.length > 0) {
    lines.push(remaining);
  }

  return lines;
}

File: crates/vorker-tui/src/theme.rs
pub const TITLE_ART: [&str; 6] = [
    "██╗   ██╗ ██████╗ ██████╗ ██╗  ██╗███████╗██████╗",
    "██║   ██║██╔═══██╗██╔══██╗██║ ██╔╝██╔════╝██╔══██╗",
    "██║   ██║██║   ██║██████╔╝█████╔╝ █████╗  ██████╔╝",
    "╚██╗ ██╔╝██║   ██║██╔══██╗██╔═██╗ ██╔══╝  ██╔══██╗",
    " ╚████╔╝ ╚██████╔╝██║  ██║██║  ██╗███████╗██║  ██║",
    "  ╚═══╝   ╚═════╝ ╚═╝  ╚═╝╚═╝  ╚═╝╚══════╝╚═╝  ╚═╝",
];

pub fn colorize(text: &str, tone: &str, enabled: bool) -> String {
    if !enabled {
        return text.to_string();
    }

    let code = match tone {
        "red" => "31",
        "green" => "32",
        "yellow" => "33",
        "brightGreen" => "92",
        "brightRed" => "91",
        "gray" => "90",
        "brightBlack" => "90",
        "magenta" => "35",
        "brightMagenta" => "95",
        "white" => "97",
        "black" => "30",
        "bgRed" => "41",
        "bgGreen" => "42",
        "bgMagenta" => "45",
        "bgGray" => "100",
        _ => "0",
    };

    format!("\u{1b}[{code}m{text}\u{1b}[0m")
}

pub fn emphasize(text: &str, enabled: bool) -> String {
    if !enabled {
        return text.to_string();
    }
    format!("\u{1b}[1m{text}\u{1b}[0m")
}

pub fn strip_ansi(text: &str) -> String {
    let mut result = String::new();
    let mut chars = text.chars().peekable();

    while let Some(ch) = chars.next() {
        if ch == '\u{1b}' && chars.peek() == Some(&'[') {
            let _ = chars.next();
            for next in chars.by_ref() {
                if next == 'm' {
                    break;
                }
            }
            continue;
        }
        result.push(ch);
    }

    result
}

pub fn visible_length(text: &str) -> usize {
    strip_ansi(text).chars().count()
}

pub fn truncate(text: &str, max_length: usize) -> String {
    if visible_length(text) <= max_length {
        return text.to_string();
    }
    if max_length <= 3 {
        return ".".repeat(max_length);
    }
    let mut output = String::new();
    for ch in strip_ansi(text).chars().take(max_length.saturating_sub(3)) {
        output.push(ch);
    }
    output.push_str("...");
    output
}

pub fn pad(text: &str, width: usize) -> String {
    let mut output = text.to_string();
    let missing = width.saturating_sub(visible_length(text));
    output.push_str(&" ".repeat(missing));
    output
}

pub fn fit(text: &str, width: usize) -> String {
    pad(&truncate(text, width), width)
}

pub fn hard_wrap(text: &str, max_length: usize) -> Vec<String> {
    let clean = strip_ansi(text);
    if clean.chars().count() <= max_length {
        return vec![clean];
    }

    let mut lines = Vec::new();
    let mut remaining = clean;

    while remaining.chars().count() > max_length {
        let slice: String = remaining.chars().take(max_length).collect();
        if let Some(break_at) = slice.rfind(' ')
            && break_at >= (max_length * 3 / 5)
        {
            lines.push(remaining[..break_at].to_string());
            remaining = remaining[break_at + 1..].to_string();
            continue;
        }
        lines.push(slice);
        remaining = remaining.chars().skip(max_length).collect();
    }

    if !remaining.is_empty() {
        lines.push(remaining);
    }

    lines
}
